### ISSUE TITLE
output a larger canvas size?

### ISSUE THREAD

**Issue by leon** [davidje13/lean-qr#4]
Hi!

Thanks for the svg implementation, it's working great :)

I was having troubles generating a larger png file though.

as you have shown in the docs:
```
code.toCanvas(canvas)
a.href = canvas.toDataURL('image/png')
```

The problem with this is that the canvas size is fixed to x number of pixels depending on the complexity of the qr code.

I was able to work around it this way

```ts
// draw the qr code to a canvas
const canvas = document.createElement('canvas')
const code = generate('mytext')
code.toCanvas(canvas)

// create a second canvas to export a larger version
const outputCanvas = document.createElement('canvas')
outputCanvas.width = 1024
outputCanvas.height = 1024
const ctx = outputCanvas.getContext('2d')
// don't smooth when we scale up
ctx.imageSmoothingEnabled = false
// first we pass in the canvas and its size, then we pass in the size we want to resize to
ctx.drawImage(
  canvas,
  0,
  0,
  canvas.width,
  canvas.height,
  0,
  0,
  1024,
  1024,
)
return outputCanvas.toDataURL('image/png')
```

I don't know if you know of an easier way to scale the canvas up, or get a larger version of the qr code?

I though it might be good to add this to the docs :)

**Comment by davidje13** [comment]
Yeah there's no built-in way to get a larger rasterised image. That's partly to keep the code small, but mostly a correctness thing (the most accurate rendering comes from setting `image-rendering: pixelated` and scaling the canvas; that ensures crisp and accurate rendering even if the screen is high-dpi or the user zooms in, whereas just rendering to a larger canvas would have blurred edges and sometimes misalignments in these situations).

But as you've found, exporting for download is perhaps an exception where it really is useful to have an upscaled image, even if it might be slightly blurry when viewed.

The code you have looks fine. In theory it should be possible to use a single canvas instead of 2 (by using `toImageData` then calling `drawImage` to render the canvas to itself) but the added complexity of handling padding probably means it isn't worth it (`toImageData` doesn't include padding, but maybe it should).

I'll have a think about whether to add this to the readme, since I don't want to encourage people to do this if they're just showing it in the browser (for the reasons I listed above), but I think it's already useful to have the code here in this issue, where people can find it if they're looking for it (along with these caveats). Maybe it would be worth adding this to the main code as a built-in "toDataURL" (so that the scaling and export as image always go hand-in-hand)

Thanks for raising this!

**Comment by davidje13** [comment]
Extracting your code as a general helper function, I could imagine something like this:

```js
function toDataURL(code, {
  type = 'image/png',
  encoderOptions = 1,
  width = null,
  height = null,
  scale = 1,
  ...options
} = {}) {
  const canvas = document.createElement('canvas');
  code.toCanvas(canvas, options);

  const outputCanvas = document.createElement('canvas');
  outputCanvas.width = width ?? (canvas.width * scale);
  outputCanvas.height = height ?? (canvas.height * scale);

  const ctx = outputCanvas.getContext('2d');
  ctx.imageSmoothingEnabled = false;
  ctx.drawImage(canvas, 0, 0, outputCanvas.width, outputCanvas.height);
  return outputCanvas.toDataURL(type, encoderOptions);
}
```

Usage:

```js
a.href = toDataURL(generate('mytext'), { type: 'image/png', width: 1024, height: 1024 });
```

**Comment by davidje13** [comment]
I added `toDataURL` to the core library ([documentation](https://github.com/davidje13/lean-qr#todataurloptions)) which should fulfil this use-case. It takes a `scale` rather than an explicit `width` / `height`, but if you really want a fixed size you can achieve it with:

```javascript
const url = code.toDataURL({ scale: 1024 / (code.size + 8) });
```

(the downside of doing this is that non-integer scales will result in the grid getting rounded to pixel boundaries, which might affect reading in some cases, especially if the resulting image gets scaled up to a larger size later)

I also added `toSvgDataURL` ([documentation](https://github.com/davidje13/lean-qr#tosvgdataurlcode-options)) for symmetry, which handles the necessary boiler-plate for wrapping SVG as a data URL.

This has been released as 1.4.0. Let me know if this works for you.

**Comment by davidje13** [comment]
since there's been no update on this I'm going to assume it's fixed and close this issue. If you're still having problems, feel free to re-open.
